### DISCUSSION TITLE
JWKS: no keys in the response of /oauth/v2/keys

### DISCUSSION THREAD

**Discussion by anatolykazantsev** [zitadel/zitadel#7198] [Q&A] [answered]
I noticed that sometimes all keys in the response of `/oauth/v2/keys`disappear (empty json array). Keys appear if I log in into the instance.

I wonder if it's a user mistake in the instance setup. I'm using instance hosted by Zitadel.

**Comment by hifabienne** [comment]
@livio-a @muhlemmer can you help here?

**Comment by muhlemmer** [comment, marked as answer]
Private keys expire after 6 hours and the corresponding  public keys 30 hours. The public are served on the keys endpoint as long as they are not yet expired. This overlap is to ensure tokens signed with an expired private key can still be  validated. The settings are managed globally in the `yaml` config file of zitadel:

https://github.com/zitadel/zitadel/blob/3d3264eb8f1420b590219dab2f4d8791302d28ae/cmd/defaults.yaml#L503-L504

https://github.com/zitadel/zitadel/blob/3d3264eb8f1420b590219dab2f4d8791302d28ae/cmd/defaults.yaml#L306-L307

After a private key expires the new keypair is only created (ad-hoc) when a new token that needs to be signed is requested. (For example an ID token or access token as JWT).

On a low traffic instance it therefore may happen that the keys endpoint returns nothing if no new (signed) tokens were created over a period of 30 hours.

**Important note**: the defaults make sure that there is sufficient overlap between token lifetime and public key lifetime. (12 vs 30 hours). If you are self-hosting and increasing token lifetime, be sure to adjust the public key lifetime accordingly or you risk not finding keys for a token that is still valid.

**Reply by muhlemmer** [reply]
As @anatolykazantsev uses zitadel cloud the last note is not of his concern. I just added that to provide a full picture of the mechanism.

> Important note: the defaults make sure that there is sufficient overlap between token lifetime and public key lifetime. (12 vs 30 hours). If you are self-hosting and increasing token lifetime, be sure to adjust the public key lifetime accordingly or you risk not finding keys for a token that is still valid.

**Reply by anatolykazantsev** [reply]
Thank you Tim!

Now it all makes sense. So my next step is to configure or modify caddy-security plugin to be compatible with such behavior. Currently the plugin tries to re-fetch public keys every time Caddy server is restarted and the server can be restarted because google restarts virtual machines to upgrade them.

Thanks, again!

**Reply by ghstahl** [reply]
guys, I get what you are doing, but that's not how the real-world works.  

please refer to this discussion.

https://github.com/zitadel/zitadel/discussions/5690

There is a reason that auth0 and keycloak do this.  

The point of the JWKS apis is that when someone calls it there MUST be public keys returned.  It doesn't matter if zitadel has signed yet or ever will.

We can't control the validators and there are many legacy systems that **POLL**, yes **POLL**, for keys.  I have seen setting where they do this every 24 hours.  

Hence that is why my settings are in the months vs hours for key rotations.  

They don't make calls to zitadel on demand if an access token comes in with a kid that they don't know about.  

We can write as many how to documents for validator libraries we want but no-one is changing them.  
i.e. nobody is listening when we tell them they need to make a n on-demand call to pull new keys.

**Reply by anatolykazantsev** [reply]
Caddy-security supports this case. There's a way to ignore missing keys on start. Later it will try another attempts to fetch them when it finds out that required key is missing in its key store.

https://github.com/greenpau/caddy-security/discussions/309

**Reply by contrun** [reply]
@anatolykazantsev I am having the same problem. Unfortunately the link you gave is gone. And both google cache and archive.org are down for the moment (google cache was gone for good).

**Reply by hifabienne** [reply]
Hi @contrun 
There was a discussion about the improvement of the key management here: https://github.com/zitadel/zitadel/discussions/7464
We have closed it now, as we have implemented it. You can find the update and information on the last comment in the discussion.
Or you can find more information here: https://zitadel.com/docs/guides/integrate/login/oidc/webkeys

**Comment by anatolykazantsev** [comment]
Hi @contrun,

You're looking for `delay_start` option in oauth identity provider settings. I couldn't find documentation for the option nor I think it existed at the moment I posted my comment. Caddy-security source code is the best documentation to it :-)

I'm using it following way in my Caddyfile:

```
{
    security {
        oauth identity provider zitadel {
            ...

            delay_start 5
            retry_attempts 3
            retry_interval 10

            ...
       }
    }
}
```
